src/Features/CSharp/Portable/Completion/CompletionProviders/SpeculativeTCompletionProvider.cs

```diff
@@ -58,46 +58,107 @@ private async Task<bool> ShouldShowSpeculativeTCompletionItemAsync(Document docu
                 return false;
             }
 
-            // If we're in a generic type argument context, use the start of the generic type name
-            // as the position for the rest of the context checks.
-            var testPosition = position;
-            var leftToken = syntaxTree.FindTokenOnLeftOfPosition(position, cancellationToken);
+            // We could be in the middle of a ref/generic/tuple type, instead of a simple T case.
+            // If we managed to walk out and get a different SpanStart, we treat it as a slightly different $$T case, otherwise it's a simple T case.
 
-            var semanticModel = await document.GetSemanticModelForNodeAsync(leftToken.Parent, cancellationToken).ConfigureAwait(false);
-            if (syntaxTree.IsGenericTypeArgumentContext(position, leftToken, cancellationToken, semanticModel))
+            var token = syntaxTree.FindTokenOnLeftOfPosition(position, cancellationToken);
+            var semanticModel = await document.GetSemanticModelForNodeAsync(token.Parent, cancellationToken).ConfigureAwait(false);
+
+            var spanStart = position;
+            while (true)
+            {
+                var orgSpanStart = spanStart;
+
+                spanStart = WalkOutOfGenericType(syntaxTree, spanStart, semanticModel, cancellationToken);
+                spanStart = WalkOutOfTupleType(syntaxTree, spanStart, cancellationToken);
+                spanStart = WalkOutOfRefType(syntaxTree, spanStart, cancellationToken);
+
+                if (spanStart == orgSpanStart)
+                {
+                    break;
+                }
+            }
```

Comment:
1. nice
2. comment something like "we weren't able to move backward, so break out so we don't loop forever"
3. `orgSpan`?  organization-span?

Reply:
😂org for original

Reply:
Renamed to `oldSpanStart`.